src/uwtools/rocoto.py

```diff
@@ -183,12 +183,26 @@ def _add_task_dependency(self, e: Element, config: dict) -> None:
         :param config: Configuration data for this element.
         """
         e = SubElement(e, STR.dependency)
+        for key in config.keys():
+            tag, _ = self._tag_name(key)
+            if tag in [STR.taskdep, STR.datadep, STR.timedep]:
+                self._add_task_dependency_operand(e, config=config)
+            else:
+                raise UWConfigError("Unhandled dependency type %s" % tag)
+
+    def _add_task_dependency_operand(self, e: Element, config: dict) -> None:
+        """
+        ???
+        """
         for key, block in config.items():
             tag, _ = self._tag_name(key)
             if tag == STR.taskdep:
                 self._set_attrs(SubElement(e, STR.taskdep), block)
+            elif tag == STR.datadep:
+                self._set_attrs(SubElement(e, STR.datadep), config={"attrs": block})
```

Comment:
@maddenp-noaa Here I added `attrs` to the beginning of the rest of the `dict` contained in `block`. I didn't update the config variable itself, but what is being passed to `_set_attrs()`. I don't think this would have an effect of the `config` being modified.

I can do as you suggested and add `deepcopy()` before calling the `_set_attrs()` method like below:
```
deepcopy(config)
self._set_attrs(SubElement(e, STR.datadep), config={"attrs": block})
```

Reply:
What you have here (I mean the code, not the comment) looks good: `block` is part of `config`, but since `_set_attrs()` doesn't modify anything, this is all read-only access and safe. It doesn't look to me like the `deepcopy()` is needed.

If this code now does what you need it do, it raises the question for me: Should we make the `datadep` YAML block have an `attrs` subtree, the same way `taskdep` apparently does? Do we have an unnecessary asymmetry here between how `taskdep` and `datadep` are represented in UW YAML? This is under our control (I think), so we should be able to make it as consistent as we like.

Just to motivate what I wrote in the first paragraph, note that a modifed subtree would be reflected in its parent object, as the whole tree is just a bunch of `dict` objects wired up together:
```
>>> config = {"a": "apple", "f": {"f1": "rose", "f2": "tulip"}, "b": "banana"}
>>> def f(config):
...   for key, value in config["flowers"].items():
...     config["flowers"][key] = value.upper()
... 
>>> f({"flowers": config["f"]})
>>> config
{'a': 'apple', 'f': {'f1': 'ROSE', 'f2': 'TULIP'}, 'b': 'banana'}
```
So, even though a new `dict` was constructed and passed to `f()`, part of that `dict`'s structure was shared with the original `config` and so, when `f()` modified that, `config` saw the modifications. Apologies if that's obvious.

Reply:
> Should we make the `datadep` YAML block have an `attrs` subtree, the same way `taskdep` apparently does? Do we have an unnecessary asymmetry here between how `taskdep` and `datadep` are represented in UW YAML?

I'm not sure if this will work as intended. Within the YAML that a user supplies to us the `datadep` tag does not have an `attrs` keyword. Only `task_*` or `workflow` tags will contain an `attrs` keyword. Similar to `datadep`, the `timedep` will also not contain an `attrs` keyword and so on with the remainder of the operators as well. I believe the asymmetry between `taskdep` and `datadep` is a necessary one due to how the Rocoto workflow is structured. 

From the [SRW documentation](https://ufs-srweather-app.readthedocs.io/en/develop/CustomizingTheWorkflow/DefineWorkflow.html#defining-dependencies) we have "`attrs`: Any of the attributes to the task tag in Rocoto. This is meant to be a subdictionary defining any of the Rocoto-supported attributes, where the key is the name of the attribute, and the value is what Rocoto expects. Attributes might include any combination of the following: cycledefs, maxtries, throttle, or final."

I believe the overlap comes into play when the sources reference the "attributes" for a tag. Although they are the attributes the keyword `attrs` will not appear in the YAML unlike how it does for the `workflow` and `taskdep` tags.

Reply:
> the keyword `attrs` will not appear in the YAML
But we get to define what appears in the YAML, right? No one is dictating that to us. We could for example decide that any `<element>` in Rocoto XML that takes attributes, e.g. `<element foo=88 bar="baz">`, will have an `attrs` subtree in YAML that contains those attributes. I'm not necessarily saying that's what we should do, but what would stand in the way of us doing it?

Reply:
I'll work on adding an `attrs` key to the `datatdep`, and `timedep` tags. I think those should be the only two we need since the operators (and, or, not, etc.) will contain one of these tags or others nested inside.